### ISSUE TITLE
Shift+Number keys not working on AC100 physical keyboard

### ISSUE THREAD

**Issue by GoogleCodeExporter** [dragonlinux/connectbot#435]
```
What steps will reproduce the problem?
1.Connect to a shell
2.Press SHIFT+1
3.Get an error beep

What is the expected output? What do you see instead?
Expect to see ! printed. Instead get an error beep.

What version of the product are you using (you can see this by using Menu
-> About in the Host List)?

ConnectBot-git-2011-02-27_11-00-43.apk on Toshiba AC100.

What type of system are you trying to connect to?

Gentoo Linux

If you are able to connect, what is the output of "echo $TERM", "uname -a",
and any other relevant information on the host?

Can't type $ :-( However, 'set' tells be TERM=screen

Linux loki 2.6.34-gentoo-r12 #3 SMP Sun Dec 5 11:28:00 GMT 2010 x86_64 AMD 
Phenom(tm) II X4 910e Processor AuthenticAMD GNU/Linux


Please provide any additional information below.

The AC100 is a smart book running Android 2.2. It doesn't have a soft keyboard, 
but does have a full(ish) physical keyboard. SHIFT works for capital letters, 
and indeed apparently all other keys on the keyboard apart from the number keys 
1-0. It even works correctly for +/= and _/-.

Connecting to a local terminal, typing SHIFT+1 gives me: ^[OP
Numbers 1-4 give me ^[O followed by P-S
Numbers 5-0 give me ^[[15~, ^[[17~, ^[[18~, ^[[19~, ^[[20~, ^[[21~
(^[[16~ is skipped for some reason).

It would also be nice if the physical CTRL keys worked (they don't appear to 
have any effect). I've tried different emulation modes (xterm and vt100) and 
this doesn't seem to have any effect on the above.  
```

Original issue reported on code.google.com by `samuel.p...@gmail.com` on 6 Mar 2011 at 2:16

**Comment by GoogleCodeExporter** [comment]
```
I have the same error with my ac100 any news about this  ?

```

Original comment by `royalmi...@gmail.com` on 21 Mar 2011 at 11:40

**Comment by GoogleCodeExporter** [comment]
```
Over here, either ALT key appears to operate as SHIFT for the number keys.
```

Original comment by `vinc...@zweije.nl` on 24 Jun 2011 at 1:53

**Comment by GoogleCodeExporter** [comment]
```
I see the same issue with the sl4a (scripting) application, in addition to 
connectbot.  I have a Motorola Xoom with a physical bluetooth keyboard, and 
after I turn on the keyboard this issue happens even when trying to input 
special characters from the soft keyboard.  Turning off the bluetooth keyboard 
at this point doesn't help.
```

Original comment by `bbla...@gmail.com` on 22 Jul 2011 at 4:41

**Comment by GoogleCodeExporter** [comment]
```
Adding on to my previous comment, I am able to get the keyboard to work 
properly if it is turned off before I start connectbot.  If I then turn on the 
keyboard, the keys such as '(' and ')' work properly.  This won't be much help 
for those with permanently attached keyboards, but maybe it will hint at the 
problem.
```

Original comment by `bbla...@gmail.com` on 22 Jul 2011 at 4:51

**Comment by GoogleCodeExporter** [comment]
```
I have the same error with my ac100 any news about this  ?

```

Original comment by `jarle...@gmail.com` on 18 Nov 2011 at 9:38

**Comment by GoogleCodeExporter** [comment]
```
Same pb here with lenovo thinkpad tablet (3.1) using lenovo's physical keyboard 
(US).
After some research: I can type '!' (on top of '1'), if I modify the 
Generic.kcm file this way:

key 1 {
    label, number:                      '1'
    base:                               '1'
    shift:                              '!'
    alt:                                '!'    # this line added
    ctrl, meta:                         none
}

... then I have to type alt+1 to get a '!' (normal behavior, given the above 
config).

So basically, it seams that ConnectBot is not sending the "SHIFT" modifier 
correctly for thoses numerical keys.
For now I'll change the charmap for keys 0, 1, 2, 3, ....
```

Original comment by `po.keept...@gmail.com` on 23 Mar 2012 at 11:02

**Comment by GoogleCodeExporter** [comment]
```
I'm experiencing this problem with a Nexus 7 and MiniSuit bluetooth keyboard.
```

Original comment by `gatoatigrado` on 11 Apr 2013 at 5:15

**Comment by GoogleCodeExporter** [comment]
```
number keys and blue keys do not work correctly
```

Original comment by `bklyn...@gmail.com` on 4 Oct 2014 at 1:32

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
Similar to in comment #6 I've used "Android Keyboard Helper" to remap
keys in a sneaky way such that almost all the physical keypresses are
the same as normal touch typing.

In short, this approach is to create a layout to set Shift keys to
Right-AltGraph, then map AltGraph+key to result in the same output as
Shift+key; with the exception of symbol keys, which are assigned to
output Symbol+null when AltGraph+SymbolKey is pressed.  

- Remapping Android keys to avoid usage of 'shift'.
- Sending Connectbot two characters per symbol (it ignores null).

This works around the the shift-lock feature of CB, and fixes symbols
at a price of introducing a null character or two symbols in a row on
a single Shift+NumKey press. (CB ignores/strips the null before it
sends on the symbol to the TERM. A sneaky hack but such a layout can
affect normal android app text fields by introducing a spurious null
character.

However, fixing symbols may require Android Keyboard Helper'. It uses
a small unicode range reserved as "triggers" to output a string in 
place of a single keypress, thus a workaround for symbols with CB.
I'm not familiar enough with the android layout/IME framework yet to
comment on how Android Keyboard Helper implements this.

I'll be adding details to Issue 459 in the future.
```

Original comment by `mtkol...@gmail.com` on 10 Oct 2014 at 6:36
